File: src/test/java/com/mercateo/jsonschema/generictype/GenericParameterizedTypeTest.java
package com.mercateo.jsonschema.generictype;

import com.googlecode.gentyref.GenericTypeReflector;
import org.junit.Test;

import java.lang.reflect.Field;
import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

public class GenericParameterizedTypeTest {

    @Test
    public void testGetContainedType() throws NoSuchFieldException {
        final Field field = TestClass.class.getDeclaredField("doubleList");
        final Type type = GenericTypeReflector.getExactFieldType(field, TestClass.class);

        @SuppressWarnings("rawtypes")
        final GenericParameterizedType<List> genericType = new GenericParameterizedType<>(
                (ParameterizedType) type, List.class);

        final GenericType<?> containedType1 = genericType.getContainedType();
        assertThat(containedType1.getRawType()).isEqualTo(List.class);

        final GenericType<?> containedType2 = containedType1.getContainedType();
        assertThat(containedType2.getRawType()).isEqualTo(Double.class);
    }

    @Test
    public void testGetContainedTypeWithGenericTypeParameter() throws NoSuchFieldException {
        final Class<?> superclass = TestClass.class.getSuperclass();
        final Field field = superclass.getDeclaredField("object");
        final Type type = GenericTypeReflector.getExactFieldType(field, TestClass.class);

        final GenericType<?> genericType = GenericType.of(type);

        assertThat(genericType.getRawType()).isEqualTo(Boolean.class);
    }

    static class SuperClass<T> {
        T object;
    }

    static class TestClass extends SuperClass<Boolean> {
        @SuppressWarnings("unused")
        List<List<Double>> doubleList;

    }

}

File: src/test/java/com/mercateo/jsonschema/generictype/GenericParameterizedTypeTest.kt
package com.mercateo.jsonschema.generictype

import com.googlecode.gentyref.GenericTypeReflector
import org.assertj.core.api.Assertions.assertThat
import org.junit.Test
import java.lang.reflect.ParameterizedType

class GenericParameterizedTypeTest {


    @Test
    fun testGetNameContainedType() {
        val field = TestClass::class.java.getDeclaredField("doubleList")
        val type = GenericTypeReflector.getExactFieldType(field, TestClass::class.java)

        @SuppressWarnings("rawtypes")
        val genericType = GenericParameterizedType(
                type as ParameterizedType, List::class.java)

        assertThat(genericType.name).isEqualTo("java.util.List<java.util.List<java.lang.Double>>")
        assertThat(genericType.simpleName).isEqualTo("java.util.List<java.util.List<java.lang.Double>>")
    }

    @Test
    fun testGetContainedType() {
        val field = TestClass::class.java.getDeclaredField("doubleList")
        val type = GenericTypeReflector.getExactFieldType(field, TestClass::class.java)

        @SuppressWarnings("rawtypes")
        val genericType = GenericParameterizedType(
                type as ParameterizedType, List::class.java)

        val containedType1 = genericType.containedType
        assertThat(containedType1.rawType).isEqualTo(java.util.List::class.java)

        val containedType2 = containedType1.containedType
        assertThat(containedType2.rawType).isEqualTo(java.lang.Double::class.java)
    }

    @Test
    fun testGetContainedTypeWithGenericTypeParameter() {
        val superclass = TestClass::class.java.superclass
        val field = superclass.getDeclaredField("object")
        val type = GenericTypeReflector.getExactFieldType(field, TestClass::class.java)

        val genericType = GenericType.of(type)

        assertThat(genericType.rawType).isEqualTo(java.lang.Boolean::class.java)
    }

    @Test
    fun supertypeShouldBeNullIfNotApplicable() {
        val field = TestClass::class.java.getDeclaredField("doubleList")
        val type = GenericTypeReflector.getExactFieldType(field, TestClass::class.java)

        val genericType = GenericType.of(type)

        val superType = genericType.superType

        assertThat(superType).isNull()
    }

}
